dist/datasource.js

```diff
@@ -98,536 +98,573 @@ function (angular, _, dateMath, moment) {
+            "intervals": intervals,
+            "limitSpec": limitSpec
+          };
+
+          if (filters && filters.length > 0) {
+            query.filter = buildFilterTree(filters, scopedVars);
+          }
 
-      if (filters && filters.length > 0) {
-        query.filter = buildFilterTree(filters, scopedVars);
-      }
+          return this._druidQuery(query);
+        };
 
-      return this._druidQuery(query);
-    };
+        this._druidQuery = function (query) {
+          var options = {
+            method: 'POST',
+            url: this.url + '/druid/v2/',
+            data: query
+          };
+          console.log("Make http request");
+          console.log(options);
+          return backendSrv.datasourceRequest(options);
+        };
 
-    this._druidQuery = function (query) {
-      var options = {
-        method: 'POST',
-        url: this.url + '/druid/v2/',
-        data: query
-      };
-      console.log("Make http request");
-      console.log(options);
-      return backendSrv.datasourceRequest(options);
-    };
+        function getLimitSpec(limitNum, orderBy) {
+          return {
+            "type": "default",
+            "limit": limitNum,
+            "columns": !orderBy? null: orderBy.map(function (col) {
+              return {"dimension": col, "direction": "DESCENDING"};
+            })
+          };
+        }
 
-    function getLimitSpec(limitNum, orderBy) {
-      return {
-        "type": "default",
-        "limit": limitNum,
-        "columns": !orderBy? null: orderBy.map(function (col) {
-          return {"dimension": col, "direction": "DESCENDING"};
-        })
-      };
-    }
-
-    function buildFilterTree(filters, scopedVars) {
-      //Do template variable replacement
-      var replacedFilters = filters.map(function (filter) {
-        return filterTemplateExpanders[filter.type](filter, scopedVars);
-      })
-      .filter(function(filter) {
-        if(filter.pattern)
-            return filter.pattern!="_REMOVE_FILTER_"
-        else if(filter.value)
-            return filter.value!="_REMOVE_FILTER_"
-        else
-            return true
+        function buildFilterTree(filters, scopedVars) {
+          //Do template variable replacement
+          var replacedFilters = filters.map(function (filter) {
+            return filterTemplateExpanders[filter.type](filter, scopedVars);
+          })
+              .map(function (filter) {
+                var finalFilter = _.omit(filter, 'negate');
+                if (filter.negate) {
+                  return { "type": "not", "field": finalFilter };
+                }
+                if (filter.type === "json") {
+                  finalFilter = JSON.parse(filter.value);
+                }
+                return finalFilter;
+              });
+
+          if (replacedFilters) {
+
+            replacedFilters = replacedFilters.filter(function (filter) {
+
+              if(filter.fields){
```

Comment:
this condition code is being repeated. we can avoid as follows
```suggestion
if (replacedFilters) {
   replacedFilters = remove_filter_recursively(replacedFilters)
}
function remove_filter_recursively(filters){
    return filters.filter(function (filter) {
        if(filter.fields) {
            remove_filter_recursively(filter.fields)
        }
        else if (filter.pattern)
            return filter.pattern !== "_REMOVE_FILTER_"
        else if (filter.value)
            return filter.value !== "_REMOVE_FILTER_"
        else if (filter.values)
            return !filter.values.includes("_REMOVE_FILTER_")
        else
            return true
    })     
}

```